src/bin/resource_tracing.rs

```diff
@@ -430,6 +441,93 @@ fn extract_pickup_data<'r>(
     }
 }
 
+fn extract_door_location<'r>(
+    scly: &structs::Scly<'r>,
+    obj: &structs::SclyObject<'r>,
+    obj_location: ScriptObjectLocation
+) -> (Option<DoorLocation>,Option<DoorLocation>) {
+
+    let scly_db = build_scly_db(scly);
+
+    let shield = search_for_scly_object(&obj.connections, &scly_db,
+        |obj| obj.property_data.as_actor()
+            .map(|sh| sh.name.to_str().unwrap().starts_with("Actor_DoorShield") &&
+                !sh.name.to_str().unwrap().contains("Key"))
+            .unwrap_or(false),
+        );
+    let unlock_shield_loc = match shield {
+        Some(shield) => Some(ScriptObjectLocation {
+            layer: scly_db[&shield.instance_id].0 as u32,
+            instance_id: shield.instance_id,
+        }),
+        None => None,
+    };
+
+    let forceunlock = search_for_scly_object(&obj.connections, &scly_db,
+        |obj| obj.property_data.as_damageable_trigger()
+            .map(|sh| sh.name.to_str().unwrap().contains("DoorUnlock"))
+            .unwrap_or(false),
+        ).unwrap();
+    let unlock_force_loc = ScriptObjectLocation {
+        layer: scly_db[&forceunlock.instance_id].0 as u32,
+        instance_id: forceunlock.instance_id,
+    };
+
+    let key_shield = search_for_scly_object(&obj.connections, &scly_db,
+        |obj| obj.property_data.as_actor()
+            .map(|sh| sh.name.to_str().unwrap().starts_with("Actor_DoorShield_Key"))
+            .unwrap_or(false),
+        );
+    let key_shield_loc = match key_shield {
+        Some(key_shield) => Some(ScriptObjectLocation {
+            layer: scly_db[&forceunlock.instance_id].0 as u32,
+            instance_id: key_shield.instance_id,
+        }),
+        None => None,
+    };
+
+    let forcekey = search_for_scly_object(&obj.connections, &scly_db,
+        |obj| obj.property_data.as_damageable_trigger()
+            .map(|sh| sh.name.to_str().unwrap().contains("DoorKey"))
+            .unwrap_or(false),
+        );
+    let key_force_loc = match forcekey {
+        Some(forcekey) => Some(ScriptObjectLocation {
+            layer: scly_db[&forcekey.instance_id].0 as u32,
+            instance_id: forcekey.instance_id,
+        }),
+        None => None,
+    };
+
+    let dock = search_for_scly_object(&obj.connections, &scly_db,
+        |obj| obj.property_data.is_dock());
+
+    let dock_number = match dock {
+        Some(dock) => Some(dock.property_data.as_dock().unwrap().dock_number),
+        // Handle the doors in Main Ventilation Shaft B
+        None => if      obj_location.instance_id == 0x150062 { Some(3) }
```

Comment:
This might be better written as

```rust
         None if obj_location.instance_id == 0x150062 => Some(3),
         None if obj_location.instance_id == 0x150066 => Some(2),
         None => None,
```